ktor-network/jvm/src/io/ktor/network/sockets/DatagramSendChannel.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.network.util.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.pool.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.selects.*
+import kotlinx.coroutines.sync.*
+import java.net.*
+import java.nio.*
+import java.nio.channels.*
+
+internal class DatagramSendChannel(
+    val channel: DatagramChannel,
+    val socket: DatagramSocketImpl
+) : SendChannel<Datagram> {
+    @ExperimentalCoroutinesApi
+    override val isClosedForSend: Boolean
+        get() = socket.isClosed
+
+    @ExperimentalCoroutinesApi
+    override val isFull: Boolean
+        get() = if (isClosedForSend) false else lock.isLocked
+
+    private val lock = Mutex()
+
+    override fun close(cause: Throwable?): Boolean {
+        if (socket.isClosed) {
+            return false
+        }
+
+        socket.close()
+        return true
+    }
```

Comment:
This does not work as expected. Try the following test (just add it to `UDPSocketTest`:
```kt
@Test
fun testClose(): Unit = runBlocking {
    val socket = aSocket(selector)
        .udp()
        .bind()
    socket.close()
}
```
It will fail with a `StackOverflowError`:
```
java.lang.StackOverflowError
	at java.lang.Class.getEnclosingMethodInfo(Class.java:1072)
	at java.lang.Class.getEnclosingClass(Class.java:1272)
	at java.lang.Class.getSimpleBinaryName(Class.java:1443)
	at java.lang.Class.getSimpleName(Class.java:1309)
	at kotlinx.coroutines.DebugStringsKt.getClassSimpleName(DebugStrings.kt:20)
	at kotlinx.coroutines.AbstractCoroutine.cancellationExceptionMessage(AbstractCoroutine.kt:97)
	at kotlinx.coroutines.JobSupport.access$cancellationExceptionMessage(JobSupport.kt:29)
	at kotlinx.coroutines.channels.ChannelCoroutine.cancel(ChannelCoroutine.kt:49)
	at kotlinx.coroutines.channels.ReceiveChannel$DefaultImpls.cancel$default(Channel.kt:311)
	at io.ktor.network.sockets.DatagramSocketImpl.close(DatagramSocketImpl.kt:48)
	at io.ktor.network.sockets.DatagramSendChannel.close(DatagramSendChannel.kt:38)
	at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:102)
	at io.ktor.network.sockets.DatagramSocketImpl.close(DatagramSocketImpl.kt:49)
	at io.ktor.network.sockets.DatagramSendChannel.close(DatagramSendChannel.kt:38)
	at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:102)
	at io.ktor.network.sockets.DatagramSocketImpl.close(DatagramSocketImpl.kt:49)
	at io.ktor.network.sockets.DatagramSendChannel.close(DatagramSendChannel.kt:38)
	at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:102)
	at io.ktor.network.sockets.DatagramSocketImpl.close(DatagramSocketImpl.kt:49)
	at io.ktor.network.sockets.DatagramSendChannel.close(DatagramSendChannel.kt:38)
	at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:102)
	at io.ktor.network.sockets.DatagramSocketImpl.close(DatagramSocketImpl.kt:49)
	at io.ktor.network.sockets.DatagramSendChannel.close(DatagramSendChannel.kt:38)
... removed duplicate lines
```

Reply:
Good catch! Thanks(and sorry for the delay). I fixed the problem in recent commits, could you take a look?

Reply:
Works great now!